### PR TITLE
QR support for UOS

### PR DESCRIPTION
**Pull request by yjkimjunior**
* [x] - QrView: would not change. However in its usage for displaying addresses, it is expected to be in `UOS introduction` format, with extra information at the end (either a `chainId` or a `genesisHash`)
* [x] - QrScanner: 
  * [x] - convert rawBytes from RNNative to U8a **(Unit Tested)**
  * [x] - strip padding, length, frame info **(Unit Tested)**
  * [x] - parse address, crypto (if substrate), payload, data as per UOS specification *(Unit Tested)*
  * [x] - encode address from bytes to ss58 **(Unit Tested)**
  * [x] - merging in from the identicon PR thank you @Tbaut for this one!
  * [x] - decode SCALE payload
    * [x] - Cold Signer SHOULD decode the transaction details from the SCALE encoding and display them to the user for verification before signing. **(Unit Tested)**
      * payload MUST be the SCALE encoding of the tuple of transaction items (nonce, call, era_description, era_header).
      * payload_hash MUST be the Blake2s 32-byte hash of the SCALE encoding of the tuple of transaction items (nonce, call, era_description, era_header).
      * immortal_payload MUST be the SCALE encoding of the tuple of transaction items (nonce, call).
    * [x] - Cold Signer SHOULD attempt to decode the message as UTF-8 encoded human readable string by whatever heuristics it finds suitable *(Unit Tested)* and display it to the user for verification before signing.
    * [x] - Cold Signer SHOULD warn the user that signing a hash is inherently insecure, in the cash of type 01. 
    * [x] - Cold Signer SHOULD (at the user's discretion) sign the message, immortal_payload, or payload if payload is of length 256 bytes or fewer. 
    * [x] - If payload is longer than 256 bytes, then it SHOULD instead sign the Blake2s hash of payload.
    * [x] - Cold Signer SHOULD display all account id values in SS58Check encoding. [#293 - It should be stored in AccountStore as such]
* [x] - ScannerStore: 
  * [x] - setDataToSign with appropriate native method based on network type, i.e. was crypto field defined (substrate) or not (eth)  **(Unit Tested)**
  * [x] - setTxToSign with appropriate hashing method based on network type. `keccak` for Ethereum, `sr25519` ~or `ed25519` for Substrate.~

### CODE REVIEW

**Comment by maciejhirsz** [comment]
I don't think this will work for you, as `@parity/qr-signer` depends on web APIs for the webcam.

File: src/components/QrView.js [hunk 1]
```diff
@@ -38,7 +47,7 @@ export default class QrView extends React.PureComponent {
   }
 
   componentDidMount() {
-    this.displayQrCode(this.props.text);
+    this.displayQrCode(this.props.accountId);
```
Comment (yjkimjunior) [inline src/components/QrView.js]:
```suggestion
    this.displayQrCode(this.props.data);
```

File: src/stores/ScannerStore.js [hunk 2]
```diff
@@ -94,16 +152,23 @@ export default class ScannerStore extends Container<ScannerState> {
...
+      const { ethereumChainId = 1 } = tx;
+      const networkKey = ethereumChainId;
+    }
+
+     // TODO cater for Substrate
     const sender = accountsStore.getById({
-      networkType: NetworkTypes.ETHEREUM,
-      networkKey,
+      protocol,
+      networkKey: networkKey || NetworkProtocols.SUBSTRATE,
```
Comment (Tbaut) [inline src/stores/ScannerStore.js]: Not sure how we should do that.
for Ethereum `networkKey === ethereumChainId`, for Substrate we could set it to `networkKey === genesisHash`? Right now it's `s + networkPrefix`.. but that doesn't help.

File: src/stores/ScannerStore.js [hunk 3]
```diff
@@ -76,13 +130,27 @@ export default class ScannerStore extends Container<ScannerState> {
   async setDataToSign(signRequest, accountsStore) {
     const message = signRequest.data.data;
     const address = signRequest.data.account;
-    const dataToSign = await ethSign(message);
+    const crypto = signRequest.data.crypto;
+    debugger;
+    let dataToSign = '';
+
+    if (crypto === 'sr25519' || crypto === 'ed25519') { // only Substrate payload has crypto field
+      const substrateSign = async () => { /* Placeholder function for now */ return Promise.resolve(message); }
```
Comment (yjkimjunior) [inline src/stores/ScannerStore.js]: for `payload: ExtrinsicPayload` -> `substrateSign(payload.raw.toHex())`
for `payloadHash: Bytes` -> `substrateSign(u8aToHex(payloadHash))`
for `message: string` -> `substrateSign(u8aToHex(new TextEncoder().encode(message)))`

File: src/stores/ScannerStore.js [hunk 4]
```diff
@@ -136,7 +213,7 @@ export default class ScannerStore extends Container<ScannerState> {
   async signData(pin = '1') {
     const { type, sender } = this.state;
     const seed = await decryptData(sender.encryptedSeed, pin);
-    const signedData = await brainWalletSign(seed, this.state.dataToSign);
+    const signedData = await brainWalletSign(seed, this.state.dataToSign); // fixme: support substrate
```
Comment (yjkimjunior) [inline src/stores/ScannerStore.js]: as above

**Comment by yjkimjunior** [comment]
Some sample QRs for testing: https://uostestqrs.herokuapp.com/ 
Source: https://github.com/yjkimjunior/UOS-QRs

I also want to add some tests for our Stores (so we don't have to manually scan the above), but `react-test-renderer` hates me and a handful of other people as well apparently -> https://github.com/facebook/react-native/issues/23943

**Comment by yjkimjunior** [comment]
Add runtime version https://github.com/polkadot-js/api/commit/a9746b3641dba784bb15deecebb417b116517909

**Comment by yjkimjunior** [comment]
### Sign Message
ezgif-2-4421c6d52e6c

### Sign Oversized Message
ezgif-2-96f04dc4aa26

### Sign Extrinsic Payload
ezgif-2-1af06b5d04e3

**Comment by yjkimjunior** [comment]
@Tbaut as mentioned before, here's where things are janky:

ezgif-2-d2aefb2aaa7e

**Comment by yjkimjunior** [comment]
Multipart will be a separate PR, though the logic is there I have no tests for them atm

**Comment by hanwencheng** [comment]
Could you please rebase the project on `master` so the commits on `master` will not be shown here? Then it will be clear to check the differences.

**Comment by yjkimjunior** [comment]
yes, doing this now

**Comment by yjkimjunior** [comment]
yes, doing this now

**Comment by yjkimjunior** [comment]
moved to #325

**Pull request closed without merging**